ros/src/twist_controller/dbw_node.py

```diff
@@ -56,49 +69,135 @@ def __init__(self):
-            #                                                     <dbw status>,
-            #                                                     <any other argument you need>)
-            if self.dbw_enabled:
-                self.publish(throttle, brake, steering)
+
+            if self._valid_state():
+                # rospy.logwarn('p.v. linear : %f, p.v. angular: %f, c.v. : %f',
+                #               self.proposed_velocities.twist.linear.x,
+                #               self.proposed_velocities.twist.angular.z,
+                #               self.current_velocity.twist.linear.x)
+
+                with self.lock:
+                    cte = compute_cte(self.waypoints, self.current_pose)
+
+                throttle, brake, steer = self.controller.control(self.activated,
+                                                                 cte,
+                                                                 self.proposed_velocities.twist.linear.x,
+                                                                 self.proposed_velocities.twist.angular.z,
+                                                                 self.current_velocity.twist.linear.x)
+                if self.activated:
+                    rospy.logwarn("%f, %f, %f", throttle, brake, steer)
+                    self.publish(throttle, brake, steer)
+
             rate.sleep()
 
     def dbw_enabled_cb(self, msg):
         self.dbw_enabled = msg.data
 
     def publish(self, throttle, brake, steer):
-        tcmd = ThrottleCmd()
-        tcmd.enable = True
-        tcmd.pedal_cmd_type = ThrottleCmd.CMD_PERCENT
-        tcmd.pedal_cmd = throttle
-        self.throttle_pub.publish(tcmd)
-
-        scmd = SteeringCmd()
-        scmd.enable = True
-        scmd.steering_wheel_angle_cmd = steer
-        self.steer_pub.publish(scmd)
-
-        bcmd = BrakeCmd()
-        bcmd.enable = True
-        bcmd.pedal_cmd_type = BrakeCmd.CMD_TORQUE
-        bcmd.pedal_cmd = brake
-        self.brake_pub.publish(bcmd)
+        if abs(throttle - self.last_throttle) < THROTTLE_EPSILON:
+            tcmd = ThrottleCmd()
+            tcmd.enable = True
+            tcmd.pedal_cmd_type = ThrottleCmd.CMD_PERCENT
+            tcmd.pedal_cmd = throttle
+            self.last_throttle = throttle
+            self.throttle_pub.publish(tcmd)
+        else:
+            rospy.logwarn("not publish throttle: %f", abs(throttle - self.last_throttle))
+
+        if abs(steer - self.last_steer) < STEERING_EPSILON:
+            scmd = SteeringCmd()
+            scmd.enable = True
+            scmd.steering_wheel_angle_cmd = steer
+            self.last_steer = steer
+            self.steer_pub.publish(scmd)
+        else:
+            rospy.logwarn("not publish steer: %f", abs(steer - self.last_steer))
+            
+
+        if abs(brake - self.last_brake) < BRAKE_EPSILON:
+            bcmd = BrakeCmd()
+            bcmd.enable = True
+            bcmd.pedal_cmd_type = BrakeCmd.CMD_TORQUE
```

Comment:
i did not look at details, but would it not be more efficient/intuitive to use PCT braking as well, rather than torque?

Reply:
We are supposed to use torque for the braking. Quoting from instructions for DBW Node:

"Note that throttle values passed to publish should be in the range 0 to 1. Brake values passed to publish should be in units of torque (N*m). The correct values for brake can be computed using the desired acceleration, weight of the vehicle, and wheel radius."

Reply:
hm, ok. but it looks like it is possible to use percentage. This is from BrakeCmd.py:

uint8 CMD_NONE=0
uint8 CMD_PEDAL=1   # Unitless, range 0.15 to 0.50
uint8 CMD_PERCENT=2 # Percent of maximum torque, range 0 to 1
uint8 CMD_TORQUE=3  # Nm, range 0 to 3250